Compile bug: Nix flake fails to build llama-server

### Git commit

834a243664114487f99520370a7a7b00fc7a486f

### Operating systems

Linux

### GGML backends

Vulkan

### Problem description & steps to reproduce

The Nix flake fails to build llama-server with the Vulkan backend on Linux, likely related to https://github.com/ggml-org/llama.cpp/issues/23055.


### First Bad Commit

253ba110bcd372207ca7b0bb56f1ea10d60d53fd

### Compile command

```shell
llama-cpp = {
    url = "github:ggml-org/llama.cpp";
    inputs.nixpkgs.follows = "nixpkgs";
};
```

### Relevant log output

```shell
llama-cpp-vulkan> [733/821] Building/provisioning WebUI assets (npm build -> HF Bucket fallback)
llama-cpp-vulkan> -- WebUI: npm not found, skipping npm build and trying HF Bucket download
llama-cpp-vulkan> -- WebUI: downloading assets from version: https://huggingface.co/buckets/ggml-org/llama-ui/resolve/b0
llama-cpp-vulkan> -- WebUI: failed to download index.html from version: "Could not resolve hostname"
llama-cpp-vulkan> -- WebUI: downloading assets from latest: https://huggingface.co/buckets/ggml-org/llama-ui/resolve/latest
llama-cpp-vulkan> -- WebUI: failed to download index.html from latest: "Could not resolve hostname"
llama-cpp-vulkan> CMake Warning at /build/source/scripts/webui-download.cmake:208 (message):
llama-cpp-vulkan>   WebUI: failed to download assets from HF Bucket (llama-ui)
llama-cpp-vulkan>
llama-cpp-vulkan>
llama-cpp-vulkan> CMake Warning at /build/source/scripts/webui-download.cmake:220 (message):
llama-cpp-vulkan>   WebUI: no source available.  Neither local build
llama-cpp-vulkan>   (/build/source/tools/server/webui) nor HF Bucket download succeeded.
llama-cpp-vulkan>
llama-cpp-vulkan>
llama-cpp-vulkan> CMake Warning at /build/source/scripts/webui-download.cmake:221 (message):
llama-cpp-vulkan>   WebUI: building server without embedded WebUI.  Set LLAMA_BUILD_WEBUI=OFF
llama-cpp-vulkan>   to suppress this warning.
llama-cpp-vulkan>
llama-cpp-vulkan>
llama-cpp-vulkan> [734/821] Generating index.html.hpp
llama-cpp-vulkan> FAILED: [code=1] tools/server/index.html.hpp /build/source/build/tools/server/index.html.hpp
llama-cpp-vulkan> cd /build/source/build/tools/server && /nix/store/q800b7nrqa8csf2gcds2i83w1icsda5w-cmake-4.1.2/bin/cmake -DINPUT=/build/source/tools/server/public/index.html -DOUTPUT=/build/source/build/
tools/server/index.html.hpp -P /build/source/scripts/xxd.cmake
llama-cpp-vulkan> CMake Error at /build/source/scripts/xxd.cmake:13 (string):
llama-cpp-vulkan>   string sub-command LENGTH requires two arguments.
```

--- a/.devops/nix/package.nix
+++ b/.devops/nix/package.nix
@@ -150,24 +150,14 @@ effectiveStdenv.mkDerivation (finalAttrs: {
       inherit nodejs;
     };
 
-    buildPhase = ''
-      runHook preBuild
-      npm run build --offline
-      runHook postBuild
-    '';
-
     installPhase = ''
-      runHook preInstall
-      mkdir -p $out
-      cp -r dist/. $out/
-      runHook postInstall
+      LLAMA_UI_OUT_DIR=$out npm run build --offline
     '';
   };
 
   postPatch = lib.optionalString useWebUi ''
-    mkdir -p build/tools/ui
-    cp -r ${finalAttrs.webui} build/tools/ui/dist
-    chmod -R u+w build/tools/ui/dist
+    cp -r ${finalAttrs.webui} tools/ui/dist
+    chmod -R u+w tools/ui/dist
   '';
 
   # With PR#6015 https://github.com/ggml-org/llama.cpp/pull/6015,

--- a/.devops/nix/package.nix
+++ b/.devops/nix/package.nix
@@ -3,6 +3,7 @@
   glibc,
   config,
   stdenv,
+  stdenvNoCC,
   runCommand,
   cmake,
   ninja,
@@ -19,6 +20,8 @@
   openssl,
   shaderc,
   spirv-headers,
+  nodejs,
+  importNpmLock,
   useBlas ?
     builtins.all (x: !x) [
       useCuda
@@ -130,7 +133,41 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     src = lib.cleanSource ../../.;
   };
 
-  postPatch = ''
+  # Builds the webui locally, taking care not to require updating any sha256 hash.
+  webui = stdenvNoCC.mkDerivation {
+    pname = "webui";
+    version = llamaVersion;
+    src = lib.cleanSource ../../tools/ui;
+
+    nativeBuildInputs = [
+      nodejs
+      importNpmLock.linkNodeModulesHook
+    ];
+
+    # no sha256 required when using buildNodeModules
+    npmDeps = importNpmLock.buildNodeModules {
+      npmRoot = ../../tools/ui;
+      inherit nodejs;
+    };
+
+    buildPhase = ''
+      runHook preBuild
+      npm run build --offline
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out
+      cp -r dist/. $out/
+      runHook postInstall
+    '';
+  };
+
+  postPatch = lib.optionalString useWebUi ''
+    mkdir -p build/tools/ui
+    cp -r ${finalAttrs.webui} build/tools/ui/dist
+    chmod -R u+w build/tools/ui/dist
   '';
 
   # With PR#6015 https://github.com/ggml-org/llama.cpp/pull/6015,

